src/dmd/cli.d

```diff
@@ -0,0 +1,383 @@
+                "help with language change identified by 'id'"
+            ),
+            Option("transition=?",
+                "list all language changes"
+            ),
+            Option("unittest",
+                "compile in unit tests"
+            ),
+            Option("v",
+                "verbose"
+            ),
+            Option("vcolumns",
+                "print character (column) numbers in diagnostics"
+            ),
+            Option("verrors=<num>",
+                "limit the number of error messages (0 means unlimited)"
+            ),
+            Option("verrors=spec",
+                "show errors from speculative compiles such as __traits(compiles,...)"
+            ),
+            Option("vgc",
+                "list all gc allocations including hidden ones"
+            ),
+            Option("vtls",
+                "list all variables going into thread local storage"
+            ),
+            Option("-version",
+                "print compiler version and exit"
+            ),
+            Option("version=<level>",
+                "compile in version code >= level"
+            ),
+            Option("version=<ident>",
+                "compile in version code identified by ident"
+            ),
+            Option("w",
+                "warnings as errors (compilation will halt)"
+            ),
+            Option("wi",
+                "warnings as messages (compilation will continue)"
+            ),
+            Option("X",
+                "generate JSON file"
+            ),
+            Option("Xf=<filename>",
+                "write JSON file to filename"
+            ),
+        ];
+    }
+}
+
+/*
+Formats the `Options` for CLI printing.
+*/
+struct CLIUsage
+{
+    /**
+        Returns a string of all available CLI options for the current targetOS.
+        Options are separated by newlines.
+    */
+    static const(char)* usage()
+    {
+        enum maxFlagLength = 18;
+        enum s = () {
+            string buf;
+            foreach (option; Usage.options)
+            {
+                if (option.os.isCurrentTargetOS)
+                {
+                    buf ~= "  -";
+                    buf ~= option.flag;
+                    // emulate current behavior of DMD
+                    if (option.flag == "defaultlib=<name>")
+                    {
+                            buf ~= "\n                    ";
+                    }
+                    else if (option.flag.length <= maxFlagLength)
+                    {
+                        foreach (i; 0 .. maxFlagLength - option.flag.length - 1)
+                            buf ~= " ";
+                    }
+                    else
+                    {
+                            buf ~= "  ";
+                    }
+                    buf ~= option.helpText;
+                    buf ~= "\n";
+                }
+            }
+            return buf;
+        }();
+        return s.ptr;
```

Comment:
Can we return a `string` and use `.ptr` at the `printf` call.  I'd like to gradually move away from using `char*`.

Reply:
Or even better not use `.ptr` at all and instead use the `printf` formatter that allows to specify pointer and length separately.

Reply:
Me too, I just used `char*` to keep the changes at a minimum. I'm more than happy to use `string` :)